### PR TITLE
Hughsie/efifb

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 does it make sense to put the `drmModeGetResources` stuff in `fu-drm-device.c` and keep the plugin simpler?

**Comment by superm1** [comment]
> @superm1 does it make sense to put the `drmModeGetResources` stuff in `fu-drm-device.c` and keep the plugin simpler?

I think so.

**Comment by hughsie** [comment]
> I think so.

Fixed.

File: libfwupdplugin/fu-drm-device.c [hunk 1]
```diff
@@ -206,9 +295,46 @@ fu_drm_device_probe(FuDevice *device, GError **error)
...
+			continue;
+		if (crtc->mode_valid && (crtc->mode.type & DRM_MODE_TYPE_PREFERRED) > 0) {
+			priv->crtc_x = crtc->x;
+			priv->crtc_y = crtc->y;
+			priv->crtc_width = crtc->width;
+			priv->crtc_height = crtc->height;
+			break;
+		}
+	}
+#endif
```
Comment (superm1) [inline libfwupdplugin/fu-drm-device.c:329]: to make this more right in more cases; what do you think about making two passes?  One that looks for eDP first, and then if no eDP found then anything with a preferred mode.

File: libfwupdplugin/fu-drm-device.c [hunk 2]
```diff
@@ -206,9 +295,46 @@ fu_drm_device_probe(FuDevice *device, GError **error)
...
+	fd = fu_io_channel_unix_get_fd(fu_udev_device_get_io_channel(FU_UDEV_DEVICE(device)));
+	res = drmModeGetResources(fd);
+	if (res == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "failed to get DRM resources");
+		return FALSE;
+	}
+	for (gint i = 0; i < res->count_crtcs; ++i) {
```
[the remark this answers is not in the document]
Comment (hughsie) [reply libfwupdplugin/fu-drm-device.c:317]: Gah. Fixed.

File: libfwupdplugin/fu-drm-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-drm-device.c:329]: I don't think eDP actually means anything special with regards to which crctc is "correct". I think the order is probably more correct.

File: plugins/uefi-capsule/fu-uefi-capsule-plugin.c [hunk 3]
```diff
@@ -1119,6 +1197,27 @@ fu_uefi_capsule_plugin_modify_config(FuPlugin *plugin,
...
+static gboolean
+fu_uefi_capsule_plugin_backend_device_added(FuPlugin *plugin,
+					    FuDevice *device,
+					    FuProgress *progress,
+					    GError **error)
+{
+	FuUefiCapsulePlugin *self = FU_UEFI_CAPSULE_PLUGIN(plugin);
+
+	/* use this DRM device to get the panel resolution */
+	if (FU_IS_DRM_DEVICE(device) && self->screen_height == 0 && self->screen_width == 0) {
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1209]: Yup, it's deliberately that as I think that's probably what the GOP driver is doing anyway. I suppose we could sort the devices my sysfs path at coldplug if you think it's needed, but this entire "just use libdrm" codepath seems like a giant workaround in all honesty.

[the remark this answers is not in the document]
Comment (superm1) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1209]: Yeah I guess you're right. This is better than nothing and probably good enough.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1209]: Nope, you're correct after-all, the order they come in is:

	/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-2
	/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-eDP-1
	/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-3
	/sys/devices/pci0000:00/0000:00:02.0/drm/card1/card1-DP-1

I think we need to sort by the `connector_id` if we want to copy the "first thing" DRM behaviour.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1209]: Nope, I'm correct after all. The only device that we successfully open is the `DRM_MINOR` type, not the connector. This means we just get `/sys/devices/pci0000:00/0000:00:02.0/drm/card1` here. That said, there's nothing stopping `card2` showing up before `card1` so I've created https://github.com/fwupd/fwupd/pull/7799 which is a pre-requisite to this one.

[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:1209]: Rebased, thanks.

**Review by superm1 — APPROVED** [review]

**Pull request merged**